### PR TITLE
feat: detect duplicate configs and warn user

### PR DESCRIPTION
**Pull request by tarcisioe**
This pull request is an attempt at duplicate configuration detection and warning as I suggested in #30.

This uses `vim.notify` with a level of `WARN`. I thought about implementing this as a health check, but the only options I could think of were:
- Having a module-level duplicates list which I could append to, and later check on `health.lua`.
- Redoing the whole lookup on `health.lua`.

I'd prefer the first one, and I don't find it too terrible since I don't think it would ever make sense to run this plugin twice. However it would be a bigger change, involves some global state, and I'd like some feedback on the idea before going down that route.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -84,6 +84,10 @@ and load it if one is found.
 [^2]: For example, a config file for a plugin called `nvim-foo` could be named `foo.lua`.
 [^3]: For example, a config file for a plugin called `foo.bar` could be named `foo-bar.lua`.
 
+If more than one is found, the first one is loaded, and
+`rocks-config` will warn the user that a duplicate configuration
+was found and skipped.
+
```
Comment (mrcjkb) [inline README.md]: suggestion: I don't think this information needs to be in the readme (as users will see the warning).

File: lua/rocks-config/init.lua [hunk 2]
```diff
@@ -33,18 +33,30 @@ local function create_plugin_heuristics(name)
...
-local function try_load_like_require(searcher, mod_name)
-    local loader = searcher(mod_name)
+---Tries to get a loader function for a given module.
+---Returns nil if the module is not found.
+---@param mod_name string The module name to search for
+---@return function | nil
+local function try_get_loader_for_module(mod_name)
+    -- Loaders that search `package.preload` and `package.path`.
+    -- We don't need to search the cpath or neovim's runtimepath,
+    -- as the nvim `lua` directory is added to the `package.path`.
```
Comment (mrcjkb) [inline lua/rocks-config/init.lua]: suggestion: this comment is no longer valid (my bad, I had forgotten to delete it)

```suggestion
```

File: lua/rocks-config/init.lua [hunk 3]
```diff
@@ -33,18 +33,30 @@ local function create_plugin_heuristics(name)
...
+        end
     end
 
+    return nil
+end
+
+---Emulates Lua's require mechanism behaviour. Lua's `require` function
+---returns `true` if the module returns nothing (`nil`), so we do the same.
+---@param loader function The loader function
+---@return any loaded
```
Comment (mrcjkb) [inline lua/rocks-config/init.lua]: issue: `any` is an EmmyLua/typescript type.
We use luaCATS.

```suggestion
---@return unknown loaded
```

File: lua/rocks-config/init.lua [hunk 4]
```diff
@@ -65,18 +77,30 @@ local function try_load_config(mod_name)
...
+
+    package.loaded[mod_name] = load_like_require(loader)
+
+    return true
+end
+
+---Checks if a plugin that already had a configuration loaded has
+---a given duplicate candidate configuration, and warns the user.
+---@param plugin_name string The plugin that is being configured
+---@param mod_name string The configuration module name to check for
```
Comment (mrcjkb) [inline lua/rocks-config/init.lua]: issue: missing `@param`
readability: I would suggest a more meaningful name than `match_name`

```suggestion
---@param plugin_name string The plugin that is being configured
---@param base_filename string The configuration's base file name
---@param mod_name string The configuration module name to check for
```

**Review by mrcjkb — CHANGES_REQUESTED** [review]
Thanks for the PR :pray: 

Looks mostly fine to me. Just some small comments.

Regarding the health check idea: I think the first option makes more sense.
But since we're actively warning the user on startup, maybe it's overkill to add a health check?

**Comment by tarcisioe** [comment]
> Regarding the health check idea: I think the first option makes more sense. But since we're actively warning the user on startup, maybe it's overkill to add a health check?

Absolutely agree. If doing it as a health check is desirable I'd remove the notify and leave everything on the health module.

File: README.md (hunk 1 above)
Comment (tarcisioe) [reply README.md]: Removed.

File: lua/rocks-config/init.lua (hunk 2 above)
Comment (tarcisioe) [reply lua/rocks-config/init.lua]: Removed.

File: lua/rocks-config/init.lua (hunk 3 above)
Comment (tarcisioe) [reply lua/rocks-config/init.lua]: Sorry, I'm not familiar with Lua type-checking. I'm using LuaLS and as long as it isn't yelling at me I assume everything is fine :sweat_smile: Fixed.

File: lua/rocks-config/init.lua (hunk 4 above)
Comment (tarcisioe) [reply lua/rocks-config/init.lua]: I changed it to `config_basename` and added the `@param`. What do you think?

**Comment by tarcisioe** [comment]
I added a commit dropping vim.notify and doing the check as a health check. I like it, works nicely. Can keep it if you guys like it, or drop it if you think notify is a better option.

File: lua/rocks-config/init.lua (hunk 3 above)
Comment (mrcjkb) [reply lua/rocks-config/init.lua]: It is a lot more lenient than I am comfortable with :sweat_smile:

File: lua/rocks-config/init.lua (hunk 4 above)
Comment (mrcjkb) [reply lua/rocks-config/init.lua]: Looks good :smile:

**Comment by mrcjkb** [comment]
> I added a commit dropping vim.notify and doing the check as a health check. I like it, works nicely. Can keep it if you guys like it, or drop it if you think notify is a better option.

Personally, I prefer the startup warning over the health check.
But I'll leave this up to @vhyrro.

**Comment by tarcisioe** [comment]
I was experimenting with adding errors while loading plugin configurations to `health.check()` and thought about a single notification like "Issues found while loading plugin configs. Run :checkhealth rocks-config for more info.".

**Comment by vhyrro** [comment]
> I was experimenting with adding errors while loading plugin configurations to `health.check()` and thought about a single notification like "Issues found while loading plugin configs. Run :checkhealth rocks-config for more info.".

I think this is the best solution. At startup we only want to check if something is wrong, not *what* is wrong, as detailed checks could slightly hinder startup time. A general check that then redirects you to the healthcheck (which can take as much time as you want in the world) is the best of both worlds.

**Comment by tarcisioe** [comment]
> At startup we only want to check if something is wrong, not _what_ is wrong, as detailed checks could slightly hinder startup time.

I agree with that as a general statement, but...

> A general check that then redirects you to the healthcheck (which can take as much time as you want in the world) is the best of both worlds.

I don't think in this specific case there is much work separation. Detecting if there is a duplicate config file is basically the same work as detecting which one it is, since we won't load it anyway. In this case I see doing it as a notification or on health check as mostly a presentation issue. However I also think that a single, general notification at startup and more detailed information on healthcheck is less overwhelming (even less annoying) to the user. Having no notification means the user may not ever notice the check, but having detailed information about the error on startup, especially when there may be a few (let's say there is duplicate configuration for two, maybe even three plugins) starts adding up on notifications.

In summary: I think we agree on the same approach, but I'd really like to point out that most of the work will be done at startup anyway (it's not really much more work than was being done before this PR).

**Comment by vhyrro** [comment]
Right, makes sense. I'm happy to go with that idea either way :)

**Comment by tarcisioe** [comment]
There it is. Also did a slight reword on the message that was added to `checkhealth` because on rereading it I found it somewhat confusing.

**Review by mrcjkb — APPROVED** [review]

**Comment by tarcisioe** [comment]
Hey, sorry to bother, is anything still missing for this to be merged? I have a following PR that I wrote on top of this and it would be easier to point it back to master if this code is already there :grimacing:

**Comment by mrcjkb** [comment]
Nothing missing from my side :smile: 
@vhyrro do you want to do the honours?

**Comment by vhyrro** [comment]
Yessir. Thank you for the PR!
